File: backend/internal/models/lobby.go
package models

type LobbyResult struct {
	Index    int    `json:"index"`
	TimeMs   int64  `json:"timeMs"`
	Scramble string `json:"scramble"`
}

type LobbyMember struct {
	User      User          `json:"user"`
	State     string        `json:"state"`
	Results   []LobbyResult `json:"results"`
	StartedAt int64         `json:"startedAt,omitempty"`
}

type Lobby struct {
	ID           int64         `json:"id"`
	HostID       int64         `json:"hostId"`
	Status       string        `json:"status"`
	AttemptIndex int           `json:"attemptIndex"`
	Scramble     string        `json:"scramble"`
	Members      []LobbyMember `json:"members"`
	CreatedAt    string        `json:"createdAt"`
	NowMs        int64         `json:"nowMs,omitempty"`
}

File: frontend/src/game/lobby.ts
import type { Attempt } from "./ao5";
import type { Lobby, LobbyMember } from "../api/lobbies";

export function lobbyMember(lobby: Lobby | null, userId: number | undefined) {
  if (!lobby || userId == null) {
    return undefined;
  }
  return lobby.members.find((member) => member.user.id === userId);
}

export function isLobbyHost(lobby: Lobby | null, userId: number | undefined) {
  return Boolean(lobby && userId != null && lobby.hostId === userId);
}

export function lobbyCanStart(lobby: Lobby | null) {
  return (lobby?.members.filter((member) => member.state === "joined").length ?? 0) > 1;
}

export function memberAttempts(member: LobbyMember | undefined): Attempt[] {
  return [...(member?.results ?? [])]
    .sort((a, b) => a.index - b.index)
    .map((result) => ({ timeMs: result.timeMs, scramble: result.scramble }));
}

export function hasAttempt(member: LobbyMember | undefined, index: number) {
  return Boolean(member?.results.some((result) => result.index === index));
}

export function lobbyInvite(lobby: Lobby | null, userId: number | undefined) {
  const me = lobbyMember(lobby, userId);
  if (!lobby || me?.state !== "invited" || lobby.status !== "open") {
    return null;
  }
  const host = lobby.members.find((member) => member.user.id === lobby.hostId)?.user;
  return { lobby, host };
}
